src/Features/Core/Portable/ImplementInterface/IImplementInterfaceService.cs

```diff
@@ -5,14 +5,56 @@
 using System.Collections.Immutable;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.CodeAnalysis.CodeActions;
 using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis.ImplementType;
 
 namespace Microsoft.CodeAnalysis.ImplementInterface;
 
+internal interface IImplementInterfaceInfo
+{
+    /// <summary>
+    /// The class or struct that is implementing the interface.
+    /// </summary>
+    INamedTypeSymbol ClassOrStructType { get; }
+
+    /// <summary>
+    /// The specific declaration node for <see cref="ClassOrStructType"/> that the interface implementations should be
+    /// added to.
+    /// </summary>
+    SyntaxNode ClassOrStructDecl { get; }
+
+    SyntaxNode InterfaceNode { get; }
+
+    /// <summary>
+    /// Set of interfaces to implement.  Normally a single interface (when a user invokes the code action on a single
+    /// entry in the interface-list for a type).  However, it may be multiple in the VB case where a user presses
+    /// 'enter' at the end of the interfaces list, where we'll implement all the missing members for all listed interfaces.
+    /// </summary>
+    ImmutableArray<INamedTypeSymbol> InterfaceTypes { get; }
+
+    ImmutableArray<(INamedTypeSymbol type, ImmutableArray<ISymbol> members)> MembersWithoutExplicitOrImplicitImplementationWhichCanBeImplicitlyImplemented { get; }
+    ImmutableArray<(INamedTypeSymbol type, ImmutableArray<ISymbol> members)> MembersWithoutExplicitOrImplicitImplementation { get; }
+    ImmutableArray<(INamedTypeSymbol type, ImmutableArray<ISymbol> members)> MembersWithoutExplicitImplementation { get; }
+}
+
+internal readonly record struct ImplementInterfaceOptions
+{
+    public bool ImplementDisposePattern { get; init; }
+    public bool Explicitly { get; init; }
+    public bool Abstractly { get; init; }
+    public bool OnlyRemaining { get; init; }
+    public ISymbol? ThroughMember { get; init; }
+}
+
 internal interface IImplementInterfaceService : ILanguageService
 {
     Task<Document> ImplementInterfaceAsync(Document document, ImplementTypeGenerationOptions options, SyntaxNode node, CancellationToken cancellationToken);
-    ImmutableArray<CodeAction> GetCodeActions(Document document, ImplementTypeGenerationOptions options, SemanticModel model, SyntaxNode node, CancellationToken cancellationToken);
+
+    Task<IImplementInterfaceInfo?> AnalyzeAsync(Document document, SyntaxNode interfaceType, CancellationToken cancellationToken);
+    Task<Document> ImplementInterfaceAsync(
+        Document document,
+        IImplementInterfaceInfo info,
+        ImplementInterfaceOptions options,
+        ImplementTypeGenerationOptions generationOptions,
+        CancellationToken cancellationToken);
```

Comment:
core part of the fix.  the service now has no knowledge of code actions.  it simply analyzes the data passed in and gives info about what it figured out.  the code action then says "based on thnat, here are hte fixes the user coudl make" and then based on that, it calls back into the service to actually make the change.